File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        vector<vector<int>> res;
        
        sort(nums.begin() , nums.end());
        
        for(int i=0;i<nums.size();i++){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            
            int a = nums[i];
            int l = i+1;
            int r = nums.size()-1;
            
            while(l<r){
                int threeSum = a + nums[l] + nums[r];
                
                if(threeSum > 0){
                    r--;
                }
                else if(threeSum < 0){
                    l++;
                }
                else{
                    res.push_back({a, nums[l], nums[r]});
                    l++;
                    while(l<r && nums[l] == nums[l-1]){
                        l++;
                    }
                }
            }
        }
        return res;
            
            
        
    }
};

/*
int main() {
    Solution solution;

    std::vector<int> nums1 = {-1, 0, 1, 2, -1, -4};
    std::vector<std::vector<int>> result1 = solution.threeSum(nums1);
    for (const auto& triplet : result1) {
        for (int num : triplet) {
            std::cout << num << " ";
        }
        std::cout << std::endl;
    }

    return 0;
}
*/

File: 0015-3sum/0015-3sum.py
## time:o(n^2)
## space: o(n)
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        ## create an ouput vector
        res = []
        
        ## sort the array
        
        nums.sort()
        
        ## now for the first value i.e a 
        
        for i , a in enumerate(nums):
            
            ## if i > 0 means first value is positve and a is equal to previous value then do nothing and move on
            
            if i > 0 and a == nums[i-1]:
                continue
                
            ## initialize the pointers for second and third values i.e b and c
            
            l , r = i+1, len(nums)-1
            while l< r:
                ## calculate the current sum
                threeSum = a + nums[l] + nums[r]
                
                if(threeSum > 0):
                    r-=1
                elif(threeSum < 0):
                    l+=1
                    
                else:
                    res.append([a, nums[l], nums[r]])
                    
                    ##update the pointers i.e only left is updated because if right pointer values are positive and might not be sumed to zero
                    l+=1
                    ## check for left pointer after increamenting that its previous value is duplicate or not
                    while l<r and nums[l] == nums[l-1]:
                        l+=1
                        
        return res           